mempool/src/pool/tx_verifier/chainstate_handle.rs

```diff
@@ -0,0 +1,243 @@
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+
+use chainstate::{
+    chainstate_interface::ChainstateInterface,
+    tx_verifier::{
+        transaction_verifier::storage::HasTxIndexDisabledError, TransactionVerifierStorageRef,
+    },
+    ChainstateError,
+};
+use chainstate_types::storage_result;
+use common::{
+    chain::{
+        tokens::{TokenAuxiliaryData, TokenId},
+        Block, DelegationId, GenBlock, OutPoint, OutPointSourceId, PoolId, Transaction,
+        TxMainChainIndex,
+    },
+    primitives::{Amount, Id},
+};
+use pos_accounting::{AccountingBlockUndo, DelegationData, PoSAccountingView, PoolData};
+use subsystem::blocking::BlockingHandle;
+use utils::shallow_clone::ShallowClone;
+use utxo::{Utxo, UtxosBlockUndo, UtxosStorageRead, UtxosView};
+
+/// Chainstate handle error includes errors coming from chainstate and inter-subprocess
+/// communication errors.
+#[derive(thiserror::Error, PartialEq, Eq, Debug)]
+pub enum Error {
+    #[error("Chainstate error: {0}")]
+    Chainstate(#[from] ChainstateError),
+    #[error("Subsystem error: {0}")]
+    Subsystem(#[from] subsystem::subsystem::CallError),
+}
+
+impl HasTxIndexDisabledError for Error {
+    fn tx_index_disabled_error() -> Self {
+        Self::Chainstate(ChainstateError::tx_index_disabled_error())
+    }
+}
+
+impl From<pos_accounting::Error> for Error {
+    fn from(e: pos_accounting::Error) -> Self {
+        Error::from(ChainstateError::from(chainstate::BlockError::from(e)))
+    }
+}
+
+impl From<utxo::Error> for Error {
+    fn from(e: utxo::Error) -> Self {
+        chainstate::tx_verifier::TransactionVerifierStorageError::from(e).into()
+    }
+}
+
+impl From<chainstate::TransactionVerifierStorageError> for Error {
+    fn from(e: chainstate::TransactionVerifierStorageError) -> Self {
+        Error::from(ChainstateError::from(chainstate::BlockError::from(e)))
+    }
+}
+
+impl From<Error> for crate::error::Error {
+    fn from(e: Error) -> Self {
+        match e {
+            Error::Chainstate(e) => crate::error::Error::TxValidationError(e.into()),
+            Error::Subsystem(_) => crate::error::Error::SubsystemFailure,
+        }
+    }
+}
+
+/// A wrapper over handle to chanistate
+pub struct ChainstateHandle<C>(BlockingHandle<C>);
+
+impl<C> ChainstateHandle<C> {
+    pub fn new(inner: subsystem::Handle<C>) -> Self {
+        Self(BlockingHandle::new(inner))
+    }
+}
+
+impl<C: 'static + Send> ChainstateHandle<C> {
```

Comment:
Too many statics all over the place... can we define a lifetime and make static an input instead of it being the default?

Reply:
I removed the generics here. I don't see this being used with multitude of parameters, and it is a bit confusing with the generics. We can re-introduce them if needed.